File: api/src/main/java/dev/obscuria/fragmentum/content/util/signal/Signal2.java
package dev.obscuria.fragmentum.content.util.signal;

import org.apache.commons.lang3.NotImplementedException;

@SuppressWarnings("unused")
public final class Signal2<P1, P2> extends Signal<Signal2.Listener<P1, P2>>
{
    public void emit(P1 p1, P2 p2)
    {
        throw new NotImplementedException();
    }

    @FunctionalInterface
    public interface Listener<P1, P2>
    {
        void consume(P1 p1, P2 p2);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal2.kt
package dev.obscuria.fragmentum.tools.signal

class Signal2<P1, P2> : Signal<Signal2.Listener<P1, P2>>() {

    fun emit(p1: P1, p2: P2) = emit { it.consume(p1, p2) }

    fun interface Listener<P1, P2> {

        fun consume(p1: P1, p2: P2)
    }
}